selfservice/strategy/saml/strategy/strategy.go

```diff
@@ -0,0 +1,352 @@
+		return nil, err
+	}
+
+	provider, err := c.Provider(c.SAMLProviders[len(c.SAMLProviders)-1].ID, c.SAMLProviders[len(c.SAMLProviders)-1].Label)
+	if err != nil {
+		return nil, err
+	}
+
+	return provider, nil
+}
+
+// Translate YAML Config file into a SAML Provider struct
+func (s *Strategy) Config(ctx context.Context) (*samlstrategy.ConfigurationCollection, error) {
+	var c samlstrategy.ConfigurationCollection
+
+	conf := s.d.Config(ctx).SelfServiceStrategy(string(s.ID())).Config
+	if err := jsonx.
+		NewStrictDecoder(bytes.NewBuffer(conf)).
+		Decode(&c); err != nil {
+		s.d.Logger().WithError(err).WithField("config", conf)
+		return nil, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to decode SAML Identity Provider configuration: %s", err))
+	}
+
+	return &c, nil
+}
+
+func (s *Strategy) populateMethod(r *http.Request, c *container.Container, message func(provider string) *text.Message) error {
+	_, err := s.Config(r.Context())
+	if err != nil {
+		return err
+	}
+
+	// does not need sorting because there is only one field
+	c.SetCSRF(s.d.GenerateCSRFToken(r))
+	// AddSamlProviders(c, conf.Providers, message)
+
+	return nil
+}
+
+func (s *Strategy) handleError(w http.ResponseWriter, r *http.Request, f flow.Flow, provider string, traits []byte, err error) error {
+	switch rf := f.(type) {
+	case *login.Flow:
+		return err
+	case *registration.Flow:
+		// Reset all nodes to not confuse users.
+		// This is kinda hacky and will probably need to be updated at some point.
+
+		rf.UI.Nodes = node.Nodes{}
+
+		// Adds the "Continue" button
+		rf.UI.SetCSRF(s.d.GenerateCSRFToken(r))
+		AddProvider(rf.UI, provider, text.NewInfoRegistrationContinue())
+
+		if traits != nil {
+			ds, err := s.d.Config(r.Context()).DefaultIdentityTraitsSchemaURL()
+			if err != nil {
+				return err
+			}
+
+			traitNodes, err := container.NodesFromJSONSchema(r.Context(), node.OpenIDConnectGroup, ds.String(), "", nil)
+			if err != nil {
+				return err
+			}
+
+			rf.UI.Nodes = append(rf.UI.Nodes, traitNodes...)
+			rf.UI.UpdateNodeValuesFromJSON(traits, "traits", node.OpenIDConnectGroup)
+		}
+
+		return err
+	}
+
+	return err
+}
+
+func (s *Strategy) CountActiveCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
+	for _, c := range cc {
+		if c.Type == s.ID() && gjson.ValidBytes(c.Config) {
+			var conf CredentialsConfig
+			if err = json.Unmarshal(c.Config, &conf); err != nil {
+				return 0, errors.WithStack(err)
+			}
+
+			for _, ider := range c.Identifiers {
+				parts := strings.Split(ider, ":")
+				if len(parts) != 2 {
+					continue
+				}
+
+				if parts[0] == conf.Providers[0].Provider && parts[1] == conf.Providers[0].Subject && len(conf.Providers[0].Subject) > 1 && len(conf.Providers[0].Provider) > 1 {
+					count++
+				}
+
+			}
+		}
+	}
+	return
+}
+
+func (s *Strategy) CountActiveFirstFactorCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
```

Comment:
This function requires a bit of reading to fully understand what it does; it might be interesting to add a comment describing what it does.